### ISSUE TITLE
How to promisify io.socket?

### ISSUE THREAD

**Issue by LeoIannacone** [balderdashy/sails.io.js#59]
Hi,

I get a error while trying to:

```
import {Promise} from 'bluebird'
import socketIOClient from 'socket.io-client'
import sailsIOClient from 'sails.io.js'

const io = sailsIOClient(socketIOClient)
Promise.promisify(io.socket.post)(server, params)
.then()
...
```

This is the error:
```
Unhandled rejection TypeError: this.request is not a function
```

Any idea how to fix it?

**Comment by cadz7** [comment]
I am trying to do the same - any update on this?

**Comment by muhammadghazali** [comment]
> This is the error:
> 
> Unhandled rejection TypeError: this.request is not a function
> Any idea how to fix it?

+1 on this issue. I got the same result when tried to promisify the `put` method.

**Comment by ksylvan** [comment]
+1 I am getting this issue too, just calling the GET (findOne) api endpoint via io.socket.get

**Comment by sailsbot** [comment]
Thanks for posting, @LeoIannacone.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by yura415** [comment]
Hey folks!
The issue is that bluebird's promisifyer works with functions that taking callback as a last argument. Callback is a function that, classically, takes first argument as error and second as some data. If first argument is not null or undefined, callback will be redirected to Promise's rejection. Sails.io's "serverResponded" callback consists of body and JWR arguments. So if some not empty body comes from Sails server, bluebird's promisifier will take that as an error and will reject the Promise.

To resolve this issue you need to specify custom promisifier, that will reject only if JWR's status code is some of the bad HTTP status code's, that are considered as error.

Example code:
```js
    function SailsIOPromisifier (originalMethod) {
      // return a function
      return function promisified () {
        var args = [].slice.call(arguments)
        // Needed so that the original method can be called with the correct receiver
        var self = this
        // which returns a promise
        return new Promise(function (resolve, reject) {
          args.push(function (body, JWR) {
            if (JWR.statusCode !== 200) {
              return reject(JWR)
            } else {
              return resolve(JWR)
            }
          })
          originalMethod.apply(self, args)
        })
      }
    }

    Promise.promisifyAll(io.socket, { promisifier: SailsIOPromisifier })

    io.socket.getAsync('/hello')
      .then(function (JWR) {
        console.log(JWR)
      })
      .catch(function (JWR) {
        console.error(JWR)
      })
```

**Comment by tylercollier** [comment]
@yura415 actually has two important parts in his answer. You can't use existing promisify solutions like Bluebird's because they assume a normal "node-style" callback, where the error is the first argument, and the `io.socket` functions don't have such a callback.

He also showed in the code the issue that the OP ran into: not putting the reference to the original object. He didn't explain it in prose, so I will here:

IF you could use Bluebird's promisify (which you can't) you'd need to do:

```
Promise.promisify(io.socket.post, io.socket)(server, params)
.then()
```

Note the second argument to `Promise.promisify`: `io.socket`. That allows the `this` to be set properly inside the `io.socket.post` method. ([Javascript's `this` is weird](http://javascriptissexy.com/understand-javascripts-this-with-clarity-and-master-it/).)

But again, you can't use `Promise.promisify` anyway, so you can use @yura415's code to do your own promisification. Below is an alternative, meant to serve as a simplification, just for `io.socket.post`. I've modified it from @yura415's above to return the body instead of the JWR, but you can should get the point:

```
const promisifiedPost = new Promise(function(resolve, reject) {
  io.socket.post(yourUrl, yourParams, function(body, jwr) {
    if (jwr.statusCode < 200 || jwr.statusCode >= 400) {
      return reject(jwr.error)
    }
    return resolve(body)
  })
})
```
